src/rust/engine/src/rule_graph.rs

```diff
@@ -251,115 +322,472 @@ impl<'t> GraphMaker<'t> {
+        all_simplified_entries.insert(entry.clone(), simplified_entries_only);
+        return Err(());
+      } else {
+        // This rule may be fulfillable, but we can't compute its complete set of dependencies until
+        // parent rule entries complete. Remove our placeholder edges before returning.
+        rule_dependency_edges.remove(&entry);
+        return Ok(ConstructGraphResult::CycledOn {
+          cyclic_deps: cycled_on,
+          partial_simplified_entries: simplified_entries_only,
+        });
+      }
+    } else {
+      // All dependencies were fulfillable and none were blocked on cycles. Remove the
+      // placeholder and store the simplified entries.
+      rule_dependency_edges.remove(&entry);
+      rule_dependency_edges.extend(simplified_entries);
+
+      all_simplified_entries.insert(entry.clone(), simplified_entries_only.clone());
+      return Ok(ConstructGraphResult::Fulfilled(simplified_entries_only));
     }
   }
 
-  fn gen_root_entries(&self, product_types: &HashSet<TypeConstraint>) -> Vec<RootEntry> {
-    let mut result: Vec<RootEntry> = Vec::new();
-    for subj_type in &self.root_subject_types {
-      for pt in product_types {
-        if let Some(entry) = self.gen_root_entry(*subj_type, pt) {
-          result.push(entry);
+  ///
+  /// Given an Entry and a mapping of all legal sources of each of its dependencies, generates a
+  /// simplified Entry for each legal combination of parameters.
+  ///
+  /// Computes the union of all parameters used by the dependencies, and then uses the powerset of
+  /// used parameters to filter the possible combinations of dependencies. If multiple choices of
+  /// dependencies are possible for any set of parameters, then the graph is ambiguous.
+  ///
+  fn monomorphize(
+    entry: &EntryWithDeps,
+    deps: Vec<(SelectKey, Vec<Entry>)>,
+  ) -> Result<HashMap<EntryWithDeps, RuleEdges>, Vec<Diagnostic>> {
+    // Collect the powerset of the union of used parameters, ordered by set size.
+    let params_powerset: Vec<Vec<TypeId>> = {
+      let mut all_used_params = BTreeSet::new();
+      for (key, inputs) in &deps {
+        for input in inputs {
+          all_used_params.extend(Self::used_params(key, input));
         }
       }
+      // Compute the powerset ordered by ascending set size.
+      let all_used_params = all_used_params.into_iter().collect::<Vec<_>>();
+      let mut param_sets = Self::powerset(&all_used_params).collect::<Vec<_>>();
+      param_sets.sort_by(|l, r| l.len().cmp(&r.len()));
+      param_sets
+    };
+
+    // Then, for the powerset of used parameters, determine which dependency combinations are
+    // satisfiable.
+    let mut combinations: HashMap<EntryWithDeps, _> = HashMap::new();
+    let mut diagnostics = Vec::new();
+    for available_params in params_powerset {
+      let available_params = available_params.into_iter().collect();
+      // If a subset of these parameters is already satisfied, skip. This has the effect of
```

Comment:
I don't follow this... Is this required? Or an optimisation?

Reply:
Did you follow up on this? Did it turn out no unification could happen, or should there be a ticket to follow up?

Reply:
I followed up, and didn't see an obvious unification. It's possible that when I dive into the tests tomorrow I can tease apart a unification, but I'm not optimistic.